src/Compilers/Core/Rebuild/CompilationOptionsReader.cs

```diff
@@ -271,35 +284,75 @@ public ImmutableArray<MetadataReferenceInfo> GetMetadataReferences()
-            foreach (var documentHandle in PdbReader.Documents.Take(sourceFileCount))
+            ImmutableArray<SyntaxTree> ResolveSyntaxTrees()
             {
-                var document = PdbReader.GetDocument(documentHandle);
-                var name = PdbReader.GetString(document.Name);
-
-                var hashAlgorithmGuid = PdbReader.GetGuid(document.HashAlgorithm);
-                var hashAlgorithm =
-                    hashAlgorithmGuid == HashAlgorithmSha1 ? SourceHashAlgorithm.Sha1
-                    : hashAlgorithmGuid == HashAlgorithmSha256 ? SourceHashAlgorithm.Sha256
-                    : SourceHashAlgorithm.None;
+                var sourceFileCount = GetSourceFileCount();
+                var builder = ImmutableArray.CreateBuilder<SyntaxTree>(sourceFileCount);
+                foreach (var tuple in GetSourceTextInfoCore())
+                {
+                    SourceText sourceText;
+                    if (ResolveEmbeddedSource(tuple.DocumentHandle, tuple.SourceTextInfo) is { } embeddedSourceTextInfo)
+                    {
+                        sourceText = embeddedSourceTextInfo.SourceText;
+                    }
+                    else
+                    {
+                        sourceText = resolver.ResolveSourceText(tuple.SourceTextInfo);
+                        if (!sourceText.GetChecksum().SequenceEqual(tuple.SourceTextInfo.Hash))
+                        {
+                            throw new InvalidOperationException();
+                        }
+                    }
 
-                var hash = PdbReader.GetBlobBytes(document.Hash);
-                var embeddedContent = ResolveEmbeddedSource(documentHandle, hashAlgorithm, encoding);
+                    var syntaxTree = createSyntaxTreeFunc(tuple.SourceTextInfo.OriginalSourceFilePath, sourceText);
+                    builder.Add(syntaxTree);
+                }
 
-                builder.Add(new SourceFileInfo(name, hashAlgorithm, hash, embeddedContent.embeddedText, embeddedContent.compressedHash));
+                return builder.MoveToImmutable();
             }
 
-            return builder.MoveToImmutable();
+            ImmutableArray<MetadataReference> ResolveMetadataReferences()
+            {
+                var builder = ImmutableArray.CreateBuilder<MetadataReference>();
+                foreach (var metadataReferenceInfo in GetMetadataReferenceInfo())
+                {
+                    var metadataReference = resolver.ResolveMetadataReference(metadataReferenceInfo);
+                    if (metadataReference.Properties.EmbedInteropTypes != metadataReferenceInfo.EmbedInteropTypes)
+                    {
+                        throw new InvalidOperationException();
+                    }
+
+                    if (!(
+                        (metadataReferenceInfo.ExternAlias is null && metadataReference.Properties.Aliases.IsDefaultOrEmpty) ||
```

Comment:
>(metadataReferenceInfo.ExternAlias is null && metadataReference.Properties.Aliases.IsDefaultOrEmpty) [](start = 24, length = 100)

Is this clause needed? (Is `Aliases.IsDefault` ever `true`? If not, it looks like the second clause may be sufficient.)

Reply:
Good catch, the implementation normalizes out the default case.

Reply:
Minor: It still looks like the first clause is unnecessary. Could the `if` condition be reduced to the following? (Sorry if I'm missing something.)
```
if (metadataReferenceInfo.ExternAlias == metadataReference.Aliases.SingleOrDefault())
{
    ...
}
```

Reply:
Ah, that's what you meant. Yeah that should work.

Reply:
Oh actually no that doesn't work. Consider the case where `externAlias` is `null` and `Aliases` has two or more items. That would false compare to `true` in your approach. 

Yes in the typical case that never happens because the compiler only creates a `MetadataReference` with zero or one items in the collection. This `MetadataReference` is being provided via the host though and they could create it with multiple items in the alias.

Reply:
> Oh actually no that doesn't work. Consider the case where externAlias is null and Aliases has two or more items. That would false compare to true in your approach. 

Is that a valid scenario? Is there also a scenario where `ExternAlias is not null` and `Aliases` has multiple items? If so, do we need to change the `SingleOrDefault()` below?